server/TriCntMsg.chpl

```diff
@@ -0,0 +1,2420 @@
+module TriCntMsg {
+
+
+  use Reflection;
+  use ServerErrors;
+  use Logging;
+  use Message;
+  use SegmentedArray;
+  use ServerErrorStrings;
+  use ServerConfig;
+  use MultiTypeSymbolTable;
+  use MultiTypeSymEntry;
+  use RandArray;
+  use IO;
+
+
+  use SymArrayDmap;
+  use Random;
+  use RadixSortLSD;
+  use Set;
+  use DistributedBag;
+  use ArgSortMsg;
+  use Time;
+  use CommAggregation;
+  use Sort;
+  use Map;
+  use DistributedDeque;
+  use GraphArray;
+
+
+  use List; 
+  use LockFreeStack;
+  use Atomics;
+  use IO.FormattedIO; 
+  use BFSMsg;
+
+
+  private config const logLevel = ServerConfig.logLevel;
+  const smLogger = new Logger(logLevel);
+  
+
+
+
+
+  // directly read a stream from given file and build a  Graph  sketch in memory
+  proc segStreamFileMsg(cmd: string, payload: string, st: borrowed SymTab): MsgTuple throws {
+      var (NeS,NvS,ColS,DirectedS, FileName,FactorS) = payload.splitMsgToTuple(6);
+      //writeln("======================Graph Reading=====================");
+      //writeln(NeS,NvS,ColS,DirectedS, FileName);
+      var Ne=NeS:int;
+      var Nv=NvS:int;
+      var Factor=FactorS:int;
+      var StreamNe=Ne/Factor:int;
+      var StreamNv=Nv/Factor:int;
+      var NumCol=ColS:int;
+
+
+      var directed=false:bool;
+      var weighted=false:bool;
+      var timer: Timer;
+      var RCMFlag=false:bool;
+      var DegreeSortFlag=false:bool;
+
+
+
+      if (DirectedS:int==1) {
+          directed=true;
+      }
+      if NumCol>2 {
+           weighted=true;
+      }
+
+      timer.start();
+      var src=makeDistArray(StreamNe,int);
+      var dst=makeDistArray(StreamNe,int);
+      //var length=makeDistArray(StreamNv,int);
+      var neighbour=makeDistArray(StreamNv,int);
+      var start_i=makeDistArray(StreamNv,int);
+
+      var e_weight = makeDistArray(StreamNe,int);
+      var v_weight = makeDistArray(StreamNv,int);
+
+      var iv=makeDistArray(StreamNe,int);
+
+      var srcR=makeDistArray(StreamNe,int);
+      var dstR=makeDistArray(StreamNe,int);
+      var neighbourR=makeDistArray(StreamNv,int);
+      var start_iR=makeDistArray(StreamNv,int);
+      ref  ivR=iv;
+
+      var linenum=0:int;
+
+      var repMsg: string;
+
+      var startpos, endpos:int;
+      var sort_flag:int;
+      var filesize:int;
+
+
+
+
+
+
+      try {
+           var f = open(FileName, iomode.r);
+           // we check if the file can be opened correctly
+           f.close();
+      } catch {
+                  smLogger.error(getModuleName(),getRoutineName(),getLineNumber(),
+                      "Open file error");
+      }
+
+
+
+      proc readLinebyLine() throws {
+           coforall loc in Locales  {
+              on loc {
+                  var randv = new RandomStream(real, here.id, false);
+                  var f = open(FileName, iomode.r);
+                  var r = f.reader(kind=ionative);
```

Comment:
Yes, this method is safer if something happens before the close file happened. I have updated the code.

Reply:
/home/z/zd4/KyleExample/remote/arkouda-njit/server/TriCntMsg.chpl:47: In function 'segStreamFileMsg':
/home/z/zd4/KyleExample/remote/arkouda-njit/server/TriCntMsg.chpl:123: error: unresolved call 'ensureClose(channel(false,native,true))'
/home/z/zd4/Mike/arkouda//src/FileIO.chpl:73: note: this candidate did not match: ensureClose(tmpf: file)
/home/z/zd4/KyleExample/remote/arkouda-njit/server/TriCntMsg.chpl:123: note: because actual argument #1 with type 'channel(false,native,true)'
/home/z/zd4/Mike/arkouda//src/FileIO.chpl:73: note: is passed to formal 'tmpf: file'

I have the above errors.

Reply:
Ah, this is my fault.  `ensureClose` only accepts a File handle and not something more generic.
I have a fix that I'm going to push to this branch and I'll log a ticket upstream to make `ensureClose` more flexible.

Reply:
I pushed a fix, if you pull this branch and run another build it should work :+1:

Reply:
Yes, it works! thanks!